### PR TITLE
LD - RequestType Service and Exception

### LINKED ISSUE #43: BACKEND: Create Service for Request Types
### Discussion

The page where new requests are created should have a dropdown menu with a variety of request types.

The request types should come from the request type table, which will be populated on application start up with the following hard coded values:

```
CS Department BS/MS program
Scholarship or Fellowship
MS program (other than CS Dept BS/MS)
PhD program
Other
```

- [x] Check for duplicates and reject any attempt to create an exact duplicate of an existing Request Type.
- [x] Application should start up with the hard-coded values

### PR DESCRIPTION
**Pull request by lodetrick**
Dependent on Issue #42 

### Description:
 - This PR creates a service for Request Type, that only saves a type if it is unique within the repository.
 - It also creates an exception, `EntityAlreadyExistsException` that is thrown by the service if the type already exists within the repository.
 - Creates the `ExampleApplicationRunner` class which saves hardcoded RequestTypes into the repository
 - Attempting to save a duplicate RequestType string will result in a BadRequest Error, returning a custom EntityAlreadyExistsException.

### Tests:
 - Needs RequestType entity and repository classes in the codebase, will rerun tests once #42 merged
 - All Tests Pass

### Issues:
 - Closes #43

### CODE REVIEW

**Review by eug-wong — DISMISSED** [review]
LGTM!

**Review by kanav9063 — DISMISSED** [review]
Changes for Request Type Serve and its tests seem correctly implemented

File: src/main/java/edu/ucsb/cs156/rec/services/RequestTypeService.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	private boolean requestTypeExists(String type, Iterable<RequestType> typeList) {
+		for (RequestType requestType : typeList) {
+			if (requestType.getRequestType().equals(type)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	public RequestType trySave(RequestType requestType) throws EntityAlreadyExistsException {
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/services/RequestTypeService.java:36]: You don't need to get ALL of the request types and iterate through them.  Let the database do the work for you.

Instead, add a method to the RequestTypeRepository called 

```java
  Optional<RequestType> findByRequestType(String requestType);
```

Just look up the RequestType and see if it exists.  If it does, it's a duplicate. If not, it isn't.

Any time you are "looping through database records" looking for something, you should stop and say: wait... there's probably a better way.

File: src/main/java/edu/ucsb/cs156/rec/services/RequestTypeService.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	public List<RequestType> trySaveTypes(List<RequestType> toSave) {
+		List<RequestType> savedTypes = new ArrayList<RequestType>();
+		Iterable<RequestType> typeList = requestTypeRepository.findAll();
+
+		toSave.forEach((requestType) -> {
+			boolean alreadyContains = requestTypeExists(requestType.getRequestType(), typeList);
+			if (!alreadyContains) {
+				savedTypes.add(requestTypeRepository.save(requestType));
+			}
+		});
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/rec/services/RequestTypeService.java:57]: This code can be greatly simplified by using 

```java
Optional<RequestType> findByRequestType(String requestType);
```

**Review by pconrad — CHANGES_REQUESTED** [review]
This is ok, but it can be made much simpler.  See comments on code review.

**Review by divyagsubramonian — APPROVED** [review]
LGTM! The code focuses on handling and storing unique request types. A new exception was created (to handle repeated request types).

**Review by pconrad — APPROVED** [review]
LGTM!

**Pull request merged**